### PR TITLE
OLS-342, OLS-347: creation and deletion of console plugin

### PR DESCRIPTION
**Pull request by raptorsun**
## Description

This PR contain the creation and deletion of the console plugin.

## Type of change

- [ ] Refactor
- [x] New feature
- [ ] Bug fix
- [ ] CVE fix
- [ ] Optimization
- [ ] Documentation Update
- [ ] Configuration Update
- [ ] Bump-up dependent library

## Related Tickets & Documents

- Related Issue #
- Closes #[OLS-347](https://issues.redhat.com//browse/OLS-347) [OLS-342](https://issues.redhat.com//browse/OLS-342)

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] PR has passed all pre-merge test jobs.
- [x] If it is a core feature, I have added thorough tests.

## Testing
- Please provide detailed steps to perform tests related to this code change.
- How were the fix/results from this change verified? Please provide relevant screenshots or results.

### CODE REVIEW

File: internal/controller/ols_console_reconciliator.go [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+			return fmt.Errorf("failed to create Console UI deployment: %w", err)
+		}
+		r.logger.Info("Console UI deployment created", "deployment", deployment.Name)
+		return nil
+	} else if err != nil {
+		return fmt.Errorf("failed to get Console UI deployment: %w", err)
+	}
+
+	if deploymentSpecEqual(&foundDeployment.Spec, &deployment.Spec) {
+		r.logger.Info("Console UI deployment unchanged", "deployment", deployment.Name)
```
Comment (vbelouso) [inline internal/controller/ols_console_reconciliator.go:136]: Same here
Throughout the code we use ".... reconciliation skipped"

File: internal/controller/ols_console_reconciliator.go [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+			return fmt.Errorf("failed to create Console Plugin: %w", err)
+		}
+		r.logger.Info("Console Plugin created", "plugin", plugin.Name)
+		return nil
+	} else if err != nil {
+		return fmt.Errorf("failed to get Console Plugin: %w", err)
+	}
+
+	if apiequality.Semantic.DeepEqual(foundPlugin.Spec, plugin.Spec) {
+		r.logger.Info("Console Plugin unchanged, skip reconciliaiton", "plugin", plugin.Name)
```
Comment (vbelouso) [inline internal/controller/ols_console_reconciliator.go:169]: Same here + typo. Must be "reconciliation"
Throughout the code we use ".... reconciliation skipped"

File: internal/controller/ols_console_reconciliator.go [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+
+	for _, task := range tasks {
+		err := task.Task(ctx, olsconfig)
+		if err != nil {
+			r.logger.Error(err, "reconcileAppServer error", "task", task.Name)
+			return fmt.Errorf("failed to %s: %w", task.Name, err)
+		}
+	}
+
+	r.logger.Info("reconcileAppServer completes")
```
Comment (vbelouso) [inline internal/controller/ols_console_reconciliator.go:53]: I guess it's typo.
Should be "ConsoleUI" or something

File: internal/controller/ols_console_reconciliator.go [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+	} else if err != nil {
+		return fmt.Errorf("failed to get Console Plugin: %w", err)
+	}
+
+	if apiequality.Semantic.DeepEqual(foundPlugin.Spec, plugin.Spec) {
+		r.logger.Info("Console Plugin unchanged, skip reconciliaiton", "plugin", plugin.Name)
+		return nil
+	}
+
+	plugin.SetResourceVersion(foundPlugin.GetResourceVersion())
```
Comment (vbelouso) [inline internal/controller/ols_console_reconciliator.go:173]: Why are we doing this?
Do we want to indicate that the resource has changed?

File: internal/controller/ols_console_reconciliator_test.go [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+var _ = Describe("Console UI reconciliator", Ordered, func() {
+
+	Context("Creation logic", Ordered, func() {
+		BeforeAll(func() {
+			console := openshiftv1.Console{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: ConsoleCRName,
+				},
+				Spec: openshiftv1.ConsoleSpec{
+					Plugins: []string{"monitoring-plugin"},
```
Comment (vbelouso) [inline internal/controller/ols_console_reconciliator_test.go:24]: lightspeed-console-plugin for th entire PR

File: internal/controller/ols_console_ui_assets.go [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+		return nil, err
+	}
+
+	return &service, nil
+}
+
+func (r *OLSConfigReconciler) generateConsoleUIDeployment(cr *olsv1alpha1.OLSConfig) (*appsv1.Deployment, error) {
+
+	replicas := int32(2)
+	val_fasle := false
```
Comment (vbelouso) [inline internal/controller/ols_console_ui_assets.go:93]: typo

File: internal/controller/ols_console_ui_assets.go [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+	val_true := true
+	volumeDefaultMode := int32(420)
```
Comment (vbelouso) [inline internal/controller/ols_console_ui_assets.go:95]: These are all default values. Do we really need to configure them?

File: internal/controller/ols_console_ui_assets.go [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+					DNSPolicy: corev1.DNSClusterFirst,
+					SecurityContext: &corev1.PodSecurityContext{
+						RunAsNonRoot: &val_true,
+						SeccompProfile: &corev1.SeccompProfile{
+							Type: "RuntimeDefault",
+						},
+					},
+				},
+			},
+			Strategy: appsv1.DeploymentStrategy{
```
Comment (vbelouso) [inline internal/controller/ols_console_ui_assets.go:189]: Same here
These are all default values.

File: internal/controller/ols_console_reconciliator.go (hunk 4 above)
Comment (raptorsun) [reply internal/controller/ols_console_reconciliator.go:173]: the k8s client refuse to update the resource in absence of resrouce version.
we can also just replace the spec of foundPlugin with generated one and use the foundPlugin to update the resource.

File: internal/controller/ols_console_reconciliator_test.go (hunk 5 above)
Comment (raptorsun) [reply internal/controller/ols_console_reconciliator_test.go:24]: This minics the Console resource in the openshift cluster, it has "monitoring-plugin" activated by default.

**Comment by raptorsun** [comment]
Thank you for the review @vbelouso  :) 
I have pushed a new commit that removes unneccesary default value settings in asset definition, added a `SetDefaults_Deployment` function to fill up the gap before comparing generated deployment with the existing one.
Please have another look.

**Comment by vbelouso** [comment]
> Thank you for the review @vbelouso :) I have pushed a new commit that removes unneccesary default value settings in asset definition, added a `SetDefaults_Deployment` function to fill up the gap before comparing generated deployment with the existing one. Please have another look.

I'm OK with that, but perhaps in the future we will need to explore additional options for comparing objects more natively.
Interesting thread https://github.com/kubernetes-sigs/kubebuilder/issues/592

**Pull request merged**